src/main/java/com/vmware/vipclient/i18n/VIPCfgFactory.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2019-2020 VMware, Inc.
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package com.vmware.vipclient.i18n;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.vmware.vipclient.i18n.exceptions.VIPClientInitException;
+
+public class VIPCfgFactory {
+    /**
+     * A lazy-loaded singleton VIPCfg using the "initialization-on-demand holder" design pattern
+     */
+    private static class VIPCfgHolder {
+        static final VIPCfg mainCfg = new VIPCfg();
+    }
+
+    public static VIPCfg getCfg() {
+        return VIPCfgHolder.mainCfg;
+    }
+
+    /**
+     * A lazy-loaded singleton Map<String, VIPCfg> configs using the "initialization-on-demand holder" design pattern
+     */
+    private static class ConfigsHolder {
+        static final Map<String, VIPCfg> configs = new HashMap<>();
+    }
+
+    public static VIPCfg getCfg(String productName) {
+        if (!contains(productName)) {
+            synchronized (ConfigsHolder.configs) {
+                if (!contains(productName)) {
+                    VIPCfg cfg = new VIPCfg();
+                    cfg.setProductName(productName);
+                }
+            }
+        }
+        return ConfigsHolder.configs.get(productName);
+    }
+
+    protected static void addCfg(VIPCfg cfg) {
+        ConfigsHolder.configs.put(cfg.getProductName(), cfg);
+    }
+
+    private static boolean contains(String productName) {
+        return ConfigsHolder.configs.containsKey(productName);
+    }
```

Comment:
Access the ConfigsHolder.configs should be synchronized and in only one location in lines from 32-40.

Reply:
Why does reading from ConfigsHolder.configs have to be synchronized? What is the thread collision use case you are trying to prevent here?

Reply:
If other threads are writing when reading, is this a problem?

Reply:
First, contains() method is private and only used inside getCfg(), so I think this is the same conversation as above.